### ISSUE TITLE
bucket event notifications creates placeholder lambda function

### ISSUE THREAD

**Issue by onedotover** [aws/aws-cdk#2781]
**Describe the bug**
When you use `function.addEventSource(new S3EventSource(...))` or `bucket.addObjectCreatedNotification(...)` it creates a BucketNotificationsHandler lambda function with some placeholder code. 

**To Reproduce**
```
    const bucket = new Bucket(this, 'TestBucket');
    const lambda = new Function(this, 'TestFunction', {
      code: Code.asset('./test-src'),
      runtime: Runtime.NodeJS10x,
      handler: 'index.handler',
      description: 'Example function',
    });
    lambda.addEventSource(new S3EventSource(bucket, {
      events: [EventType.ObjectCreated],
    }));
```
**Expected behavior**
Should only connect the TestFunction to the lambda function not create a new placeholder function.

**Version:**
 - macOS 10.14.5
 - Typescript
 - 0.33

**Comment by mirskiy** [comment]
Running in to the same issue. Tried downgrading to 0.29, didn't help.

**Comment by made2591** [comment]
I was able to reproduce the error. If it's ok for anyone, I can work on this (I'm investigated already a bit)

**Comment by made2591** [comment]
Ok I think I found the reason of this problem, and it's described [here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-notificationconfig.html). Actually, what CDK (version 0.35) implements, is discussed also [here](https://aws.amazon.com/blogs/mt/resolving-circular-dependency-in-provisioning-of-amazon-s3-buckets-with-aws-lambda-event-notifications/).

If you create the target resource and related permissions in the same template, you might have a circular dependency. For example, you might use the `AWS::Lambda::Permission` resource to grant the bucket permission to invoke an AWS Lambda function. However, AWS CloudFormation can't create the bucket until the bucket has permission to invoke the function (AWS CloudFormation checks whether the bucket can invoke the function). If you're using Refs to pass the bucket name, this leads to a circular dependency.

This is because `AWS::Lambda::Permission` needs to know the ARN of the `AWS::S3::Bucket` before it can create `AWS::Lambda::Function`. However, the `AWS::S3::Bucket`'s `NotificationConfiguration` property needs to know the ARN of `AWS::Lambda::Function` before it can create the `AWS::S3::Bucket`, thus creating a circular dependency. The solution discussed and implemented avoid the circular dependency demonstrated earlier by creating the S3 bucket without any notification configuration. Then it sets up the notification after the Lambda function, Lambda permission, and S3 bucket resources have been created.

In the code, you can follow the whole process starting [here](https://github.com/awslabs/aws-cdk/blob/ea10f0d141a48819ec0000cd7905feda993870a9/packages/%40aws-cdk/aws-s3/lib/notifications-resource/notifications-resource.ts#L52): the `createResourceOnce()` call the `NotificationsResourceHandler.singleton(this);`. Then, a stack-singleton lambda function with the logic for a CloudFormation custom resource provisions bucket notification configuration for a bucket with a well-known logicalID ([source](https://github.com/awslabs/aws-cdk/blob/ea10f0d141a48819ec0000cd7905feda993870a9/packages/%40aws-cdk/aws-s3/lib/notifications-resource/notifications-resource-handler.ts#L29)). The line 35 verifies that a Lambda with that ID doesn't exist yet - it will *never* exist. Then, the bucket notifications resources in the stack are defined only once and lazily, so that if notifications are not added, there is no notifications resource.

To conclude... I don't think there's a way to do avoid this (@eladb @rix0rrr am I wrong?), excluding having two separate Stacks that follow the same reasoning - thus, by defining the two resource (exposing the Lambda from one Stack to another). Not sure if it could work, I will try and let you know asap.

Hope to be helpful!

**Comment by made2591** [comment]
@NGL321 what do you think about [this](https://github.com/awslabs/aws-cdk/issues/2781#issuecomment-503766858)?

**Comment by NGL321** [comment]
@made2591 I will take a little more time to read through what you have written later this afternoon (thank you for being so thorough!), but in the interim, if you feel that you have a fix, by all means go ahead and work on it!

**Comment by NGL321** [comment]
@made2591 I will take a little more time to review your research either this evening or Monday morning (thank you for being so thorough!).
In the meantime, if you think you can fix it yourself, please feel free to work on the issue!

**Comment by jthornbrue** [comment]
My stack contains an object created notification that triggers an SNS topic. See below (using the Python API):

my_bucket.add_event_notification(
        aws_s3.EventType.OBJECT_CREATED,
        aws_s3_notifications.SnsDestination(my_topic),
        aws_s3.NotificationKeyFilter(
            prefix=OUTPUT_PREFIX,
            suffix='.csv'
        )
    )

This creates a dummy lambda function in the stack--even though there is not and never will be a lambda target. This is annoying. Any progress on this bug?

**Comment by LechuckThePirate** [comment]
This should be solved. I don't understand how this went out of beta with this problem. If you run into this, it's not just that you have a lambda function there that shouldn't be there... the thing is you are actually PAYING for every S3 bucket event...

**Comment by erikaadvisser** [comment]
For me this problem seems even worse. Along with the placeholder lambda, also a placeholder role and policy are created, as cdk diff shows:
...
[+] AWS::IAM::Role BucketNotificationsHandler050a0587b7544547bf325f094a3db834/Role BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC
[+] AWS::IAM::Policy BucketNotificationsHandler050a0587b7544547bf325f094a3db834/Role/DefaultPolicy BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36
[+] AWS::Lambda::Function BucketNotificationsHandler050a0587b7544547bf325f094a3db834 BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691

My company enforces that all roles have a permission boundary, and the placeholder role does not have one. So my entire stack fails. I see no way to get around this (but I am new to CDK)

**Comment by Roman-L** [comment]
Are there any workarounds for this?

@made2591 Did you test the multiple stacks solution?

**Comment by dz902** [comment]
Circular reference is annoying, but could be avoided with specifying ARNS instead of using refs. At least we should have an option to avoid the extra Lambda and document the behavior.

Although I see that extra Lambda was not invoked by anything anywhere IF you bind the event to another Lambda.

**Comment by nija-at** [comment]
> the thing is you are actually PAYING for every S3 bucket event...

@LechuckThePirate - how do you mean? As far as I can tell, we're adding the additional lambda as a `CustomResource` on the CloudFormation stack. This lambda is invoked once per deployment and not once per S3 event. 

> My company enforces that all roles have a permission boundary, and the placeholder role does not have one. So my entire stack fails. I see no way to get around this (but I am new to CDK)

@erikaadvisser - [Here is the additional policy](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-s3-notifications/test/integ.notifications.expected.json#L188-L208) that is applied to the bucket notification lambda function. Is `"Resource": "*"` the bit that restricted by your company?

**Comment by erikaadvisser** [comment]
@nija-at , I think that it is the creation of the IAM:Role part that failed, not the creation of the policy. The role needs a permission boundary. However, I no longer work for this company, so I cannot check it for sure.

**Comment by nija-at** [comment]
Opened this issue to track the permission boundary - https://github.com/aws/aws-cdk/issues/5925

Unfortunately, the placeholder lambda is unavoidable. See [made2591's comment](https://github.com/aws/aws-cdk/issues/2781#issuecomment-503766858) for details.

I don't see any other specific issue reported here. Feel free to reopen if I missed something something or a separate issue if it hasn't been reported here.

**Comment by dballance** [comment]
@nija-at - I'm new to the CDK - but ran into this in my first experimentation. Some googling led me to this page. 

How is the CDK deployment different from the SAM CFN deployments? I only ask because via a SAM template, I can deploy both a bucket resource and establish my lambda event without triggering a circular dependency.

```
Resources:
  Lambda:
    Type: AWS::Serverless::Function
    Properties: 
      Events: 
        SomeS3Event:
          Type: S3
          Properties:
            Bucket: !Ref SourceBucket
            Events: s3.ObjectCreated:*
  SourceBucket:
    Type: AWS::S3::Bucket
    Properties:
      BucketName: test-bucket
```

**Comment by jaidisido** [comment]
This workaround is a bit too opinionated and not flexible enough in my view. One major limitation we are encountering is that we have no control whatsoever on the created Lambda. For instance, every Lambda needs to be within a VPC and encrypted in our prod environments and I cannot see how we can apply this to this auto-generated Lambda

**Comment by falkvonohlen** [comment]
> This workaround is a bit too opinionated and not flexible enough in my view. One major limitation we are encountering is that we have no control whatsoever on the created Lambda. For instance, every Lambda needs to be within a VPC and encrypted in our prod environments and I cannot see how we can apply this to this auto-generated Lambda

We have the same issue for our AWS accounts. For that reason I created a separate issue and a pull request to add a new bucket attribute notificationsHandlerVpc. 
Issue: #20903
Pull Request: #20904

**Comment by KalebRoncatti-ciandt** [comment]
Do we have a solution for that in the current version of CDK ? whenever I create a lambda stack with an event notification (using `v2.31`) based on a Prefix (using Python) a dummy notification lambda is also created.

**Comment by MW-Hochmann** [comment]
Is there any working workaround or at least a way to change the python version with the lambda which gets created this way? My company is still on cdkv2 and when I try to use addEventNotification cdk fails with the error:

`"The runtime parameter of python3.7 is no longer supported for creating or updating AWS Lambda functions. We recommend you use a supported runtime while creating or updating functions. `

I don't care that it creates a Lamdba I don't need, but it seems that this Lambda cannot be created so cdk always fails.
